### PR TITLE
fix: patch additions honor source key style

### LINKED ISSUE #4928: kustomize fails to apply overlay if key in base yaml is numeric and missing
### What happened?

kustomize failed to apply an overlay.

More details in the "How can we reproduce" section

### What did you expect to happen?

I expected it to apply the overlay correctly, by overriding the key that is already defined in the base and adding the key that isn't present in the base.yaml.

### How can we reproduce it (as minimally and precisely as possible)?

This bug has occurred in a real production scenario, here follows some instruction to reproduce and test the issue.

### filesystem tree
```
.
├── base.yaml
├── kustomization.yaml
├── overlays
│   └── overlay.yaml
└── test-kustomize-versions.sh
```

### files

base.yaml
```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: blabla
  namespace: blabla-ns
data:
  "6443": "foobar"
```

overlays/overlay.yaml
```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: blabla
  namespace: blabla-ns
data:
  "6443": "barfoo"
  "9110": "foo-foo"
```

kustomization.yaml
```yaml
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization  
bases:
  - base.yaml
patchesStrategicMerge:
  - overlays/overlay.yaml
```

test-kustomize-versions.sh
```bash
#!/usr/bin/env bash

echo "---------------------------------------------------------"

for version in $(asdf list kustomize | tr -d ' '); do
	asdf local kustomize "$version"
	echo "Running kustomize $(kustomize version --short)"

	kustomize build && echo "SUCCESS" || echo "FAILED"
	echo "---------------------------------------------------------"
	sleep 1
done
```

### Expected output

```yaml
apiVersion: v1
data:
  "6443": barfoo
  "9110": "foo-foo"
kind: ConfigMap
metadata:
  name: blabla
  namespace: blabla-ns
```

### Actual output

```
./test-kustomize-versions.sh

---------------------------------------------------------
Running kustomize {kustomize/v4.2.0  2021-06-30T22:49:26Z  }
Error: map[string]interface {}{"apiVersion":"v1", "data":map[interface {}]interface {}{9110:"foo-foo", "6443":"barfoo"}, "kind":"ConfigMap", "metadata":map[string]interface {}{"name":"blabla", "namespace":"blabla-ns"}}: json: unsupported type: map[interface {}]interface {}
FAILED
---------------------------------------------------------
Running kustomize {kustomize/v4.3.0  2021-08-24T19:24:28Z  }
Error: map[string]interface {}{"apiVersion":"v1", "data":map[interface {}]interface {}{9110:"foo-foo", "6443":"barfoo"}, "kind":"ConfigMap", "metadata":map[string]interface {}{"name":"blabla", "namespace":"blabla-ns"}}: json: unsupported type: map[interface {}]interface {}
FAILED
---------------------------------------------------------
Running kustomize {kustomize/v4.4.1  2021-11-11T23:36:27Z  }
Error: map[string]interface {}{"apiVersion":"v1", "data":map[interface {}]interface {}{9110:"foo-foo", "6443":"barfoo"}, "kind":"ConfigMap", "metadata":map[string]interface {}{"name":"blabla", "namespace":"blabla-ns"}}: json: unsupported type: map[interface {}]interface {}
FAILED
---------------------------------------------------------
Running kustomize {kustomize/v4.5.0  2022-02-02T00:20:43Z  }
Error: map[string]interface {}{"apiVersion":"v1", "data":map[interface {}]interface {}{9110:"foo-foo", "6443":"barfoo"}, "kind":"ConfigMap", "metadata":map[string]interface {}{"name":"blabla", "namespace":"blabla-ns"}}: json: unsupported type: map[interface {}]interface {}
FAILED
---------------------------------------------------------
Running kustomize {kustomize/v4.5.7  2022-08-02T16:35:54Z  }
Error: map[string]interface {}{"apiVersion":"v1", "data":map[interface {}]interface {}{9110:"foo-foo", "6443":"barfoo"}, "kind":"ConfigMap", "metadata":map[string]interface {}{"name":"blabla", "namespace":"blabla-ns"}}: json: unsupported type: map[interface {}]interface {}
FAILED
---------------------------------------------------------
```

### Kustomize version

4.5.7

### Operating system

None

### PR DESCRIPTION
**Pull request by ephesused**
When a patch appends a new node, it should honor the key style from the patch. Prior to this commit, no style was applied, leading to problems when the key value could be interpreted as a different type based on its content. For example, "9110" needs quoting to ensure it is seen as a string in yaml.

Fixes #4928

This PR replaces #5005, where I messed up my working branch. Given circumstances, I figured it was simpler to just cut a new PR.

### CODE REVIEW

File: api/filters/patchstrategicmerge/patchstrategicmerge_test.go [hunk 1]
```diff
@@ -776,6 +776,103 @@ spec:
...
+		// Issue #4928
+		"support numeric keys": {
+			input: `
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: blabla
+  namespace: blabla-ns
+data:
+  "6443": "foobar"
```
Comment (koba1t) [inline api/filters/patchstrategicmerge/patchstrategicmerge_test.go:791]: Could you consider adding a test case when the key-value has `"6443"` and `6443`?

Comment (koba1t) [reply api/filters/patchstrategicmerge/patchstrategicmerge_test.go:791]: I'm sorry for writing a misreadable comment.
I just want to add to tests that mixed the same name string key and int key.

Ex.
- base configMap has `"6443": "foobar"` and a patch has `6443: "barfoo_int"`

Comment (ephesused) [reply api/filters/patchstrategicmerge/patchstrategicmerge_test.go:791]: Done - note that this scenario mirrors one that I discussed with @KnVerey in my earlier attempt: https://github.com/kubernetes-sigs/kustomize/pull/5005/commits/4a01fd33de5b51c05c3053df2ae284e1683f8b00#r1156522627

A key point to make is that I'm intending this fix to be targeted at the specific issue in #4928: the key style of a node that's newly created in a merge.

This PR is not intended to alter the existing key style behavior for patches (which, I admit, is somewhat confusing). In that context, the adjusted test case we are discussing falls outside of the scope of what this PR changes. The current behavior for kustomize when input and patch have a matching key is to retain the key style of the input. This PR does not change that behavior.

Comment (koba1t) [reply api/filters/patchstrategicmerge/patchstrategicmerge_test.go:791]: > In that context, the adjusted test case we are discussing falls outside of the scope of what this PR changes.

I understood this sentence. I want to keep until behavior with test cases.

Comment (koba1t) [reply api/filters/patchstrategicmerge/patchstrategicmerge_test.go:791]: So I want to say that we need to add test cases while keeping at the tests from https://github.com/kubernetes-sigs/kustomize/pull/5196/commits/691b7d1df33d9ead7c37dde763b2a4200e24bada.
Could you add test cases using the int key under your tests?

Comment (ephesused) [reply api/filters/patchstrategicmerge/patchstrategicmerge_test.go:791]: Apologies - I believe I've got it right this time. Thanks for your patience.

**Comment by Hammond95** [comment]
any updates on this?

**Comment by ephesused** [comment]
> Like the other PR, would you mind pushing a dummy commit to see if it retriggers the tests?

I simply merged to make this PR current. I see the workflows waiting to run, pending approval from a maintainer. Thanks!

**Review by natasha41575 — APPROVED** [review]

**Pull request merged**